File: 2022/clojure/p02.clj
;; # Day 2
(ns p02
  (:require [clojure.string :as str]))

(defn parse [s]
  (map keyword (str/split s #" ")))

(def data
  (->> (str/split-lines (slurp "../input/02.txt"))
       (map parse)))

(def lookup
  {:A :rock
   :B :paper
   :C :scissors
   :X :rock
   :Y :paper
   :Z :scissors})

(def beat
  {:rock :scissors
   :scissors :paper
   :paper :rock})

;; Each choice has a shape-score.
(def scores
  {:rock 1
   :paper 2
   :scissors 3})

(defn score
  "The score for a round is the sum of the score for the shape and the outcome."
  [[a b]]
  (+ (scores b)
     (cond
       ;; lose
       (= b (beat a)) 0
       ;; tie
       (= a b) 3
       ;; win
       :else 6)))

(def ans1
  (transduce
   (comp
    (map #(map lookup %))
    (map score))
   +
   data))

(defn complete [[a b]]
  (let [a (lookup a)]
   (case b
     :X [a (beat a)]
     :Y [a a]
     :Z [a (beat (beat a))])))

(def ans2
  (transduce
   (comp
    (map complete)
    (map score))
   +
   data))


(defn -main [& _]
  (println "Answer1:" ans1)
  (println "Answer2:" ans2))

File: 2022/python/p02.py
beats = {"rock": "scissors", "paper": "rock", "scissors": "paper"}

lookup = {
    "A": "rock",
    "B": "paper",
    "C": "scissors",
    "X": "rock",
    "Y": "paper",
    "Z": "scissors",
}

shape = {"rock": 1, "paper": 2, "scissors": 3}


def score(game):
    opponent, you = game
    shape_score = shape[you]
    if opponent == you:
        # draw
        return shape_score + 3
    elif beats[opponent] == you:
        # you lost
        return shape_score
    else:
        # you won
        return shape_score + 6


data = []
with open("../input/02.txt") as f:
    for line in f:
        data.append(tuple(line.strip().split()))


part1 = sum(score(map(lookup.get, game)) for game in data)
print(f"Answer1: {part1}")

loses_to = {v: k for k, v in beats.items()}


def part2_score(game):
    first, requirement = game
    opponent = lookup[first]
    if requirement == "X":
        # must lose
        you = beats[opponent]
    elif requirement == "Y":
        # must draw
        you = opponent
    else:
        # must win
        you = loses_to[opponent]
    return score((opponent, you))


part2 = sum(part2_score(game) for game in data)
print(f"Answer2: {part2}")
